clang/lib/DPCT/ASTTraversal.cpp

```diff
@@ -8423,6 +8426,86 @@ void KernelCallRule::registerMatcher(ast_matchers::MatchFinder &MF) {
+  const auto &SM = DpctGlobalInfo::getSourceManager();
+  auto KCallSpellingRange = getTheLastCompleteImmediateRange(
+      KCall->getBeginLoc(), KCall->getEndLoc());
+  auto LocInfo = dpct::DpctGlobalInfo::getLocInfo(KCallSpellingRange.first);
+  std::string FilePath = llvm::sys::path::convert_to_slash(
+      StringRef(DpctGlobalInfo::removeSymlinks(
+          SM.getFileManager(), LocInfo.first.getCanonicalPath().str())));
+  auto InRootPath = llvm::sys::path::convert_to_slash(
+      DpctGlobalInfo::getInRoot().getCanonicalPath());
+  size_t FilePathCount = std::count_if(FilePath.begin(), FilePath.end(),
+                                       [](char c) { return c == '/'; });
+  size_t InRootPathCount = std::count_if(InRootPath.begin(), InRootPath.end(),
+                                         [](char c) { return c == '/'; });
+
+  llvm::SmallString<512> RelativePath;
+
+  std::string DebugArgsString = "(\"";
+  std::string DebugArgsStringSYCL = "(\"";
+  DebugArgsString += KCallSpellingRange.first.printToString(SM) + "\", ";
+  DebugArgsStringSYCL +=
+      KCallSpellingRange.first.printToString(SM) + "(SYCL)\", ";
+  std::string StramStr = "0";
+  if (auto *Config = KCall->getConfig()) {
+    if (Config->getNumArgs() > 3) {
+      auto StramStrSpell = getStmtSpelling(Config->getArg(3));
+      if (StramStrSpell.compare("")) {
+        StramStr = StramStrSpell;
+      }
+    }
+  }
+  int Index = getPlaceholderIdx(KCall);
+  if (Index == 0) {
+    Index = DpctGlobalInfo::getHelperFuncReplInfoIndexThenInc();
+  }
+  buildTempVariableMap(Index, KCall, HelperFuncType::HFT_DefaultQueue);
+  std::string QueueStr = "{{NEEDREPLACEQ" + std::to_string(Index) + "}}";
+  DebugArgsString += StramStr;
+  DebugArgsStringSYCL += QueueStr;
+  for (auto *Arg : KCall->arguments()) {
+    if (const auto *DRE = dyn_cast<DeclRefExpr>(Arg->IgnoreImpCasts())) {
+      DebugArgsString += ", ";
+      DebugArgsStringSYCL += ", ";
+      std::string SchemaStr = DpctGlobalInfo::getVarSchema(DRE);
+      DebugArgsString += SchemaStr + ", ";
+      DebugArgsStringSYCL += SchemaStr + ", ";
+      DebugArgsString += "(long *)&" + getStmtSpelling(Arg);
+      DebugArgsStringSYCL += "(long *)&" + getStmtSpelling(Arg);
+    }
+  }
+  DebugArgsString += ");" + std::string(getNL());
+  DebugArgsStringSYCL += ");" + std::string(getNL());
+  emplaceTransformation(new InsertText(
+      KCallSpellingRange.first,
+      "dpct::experimental::gen_prolog_API_CP" + DebugArgsString, 0, true));
+  emplaceTransformation(new InsertText(
+      KCallSpellingRange.first,
+      "dpct::experimental::gen_prolog_API_CP" + DebugArgsStringSYCL, 0, false));
+  emplaceTransformation(new InsertText(
+      EpilogLocation, "dpct::experimental::gen_epilog_API_CP" + DebugArgsString,
+      0, true));
+  emplaceTransformation(new InsertText(
+      EpilogLocation,
+      "dpct::experimental::gen_epilog_API_CP" + DebugArgsStringSYCL, 0, false));
+  std::string SchemaRelativePath = "";
```

Comment:
It is better to add assert to detect overflow of "FilePathCount - InRootPathCount"